### PR TITLE
fix(dialog): allow easier dialog control through the dialog service

### LINKED ISSUE #745: Icon tooltip on top of actionable element Issue
Hi Team,

My use case is that I have to add icon tooltip in **table header cell which is also sortable**. So, when I click on icon it opens tooltip but also sorts the table. To fix this issue we got a feedback from carbon team to use stopPropagation on icon click. This fixed the issue but because of this fix, other tooltip on the page which is already open doesn't automatically close. 

I am using below carbon version:
"carbon-components": "^10.5.1",
Screenshot 2019-09-12 at 6 02 10 PM
"carbon-components-angular": "^3.14.4",

### PR DESCRIPTION
**Pull request by cal-smith**
closes #745

### CODE REVIEW

File: src/dialog/dialog.service.ts [hunk 1]
```diff
@@ -80,59 +107,106 @@ export class DialogService {
...
+		if (this.dialogSubscription) {
+			this.dialogSubscription.unsubscribe();
+		}
+	}
+
+	/**
+	 * Closes all known `Dialog`s. Does not focus any previous elements, since we can't know which would be correct
+	 */
+	closeAll() {
+		DialogService.dialogRefs.forEach(ref => {
```
Comment (zvonimirfras) [inline src/dialog/dialog.service.ts]: Having it in one line, no curlies, could improve readability.

File: src/dialog/dialog.service.ts [hunk 2]
```diff
@@ -80,59 +107,106 @@ export class DialogService {
...
+			});
+		} else {
+			dialogRef = this.placeholderService.createComponent(componentFactory, this.injector);
+		}
+
+		// keep track of all initialized dialogs
+		DialogService.dialogRefs.add(dialogRef);
+
+		// initialize some extra options
+		let focusedElement = document.activeElement;
```
[the remark this answers is not in the document]
Comment (cal-smith) [reply src/dialog/dialog.service.ts]: Yep, I think that was left over from the previous, more complex, logic

File: src/dialog/dialog.service.ts [hunk 3]
```diff
@@ -80,59 +107,106 @@ export class DialogService {
...
+		DialogService.dialogCloseSubscription.add(closeSubscription);
 
-			this.dialogSubscription = this.onClose.subscribe(() => {
-				this.close(viewContainer);
-			});
+		dialogRef.instance.elementRef.nativeElement.focus();
 
-			this.dialogRef.instance.elementRef.nativeElement.focus();
-		}
+		// deprecated - kept for API compat
```
[the remark this answers is not in the document]
Comment (cal-smith) [reply src/dialog/dialog.service.ts]: fair 😆

File: src/dialog/dialog.service.ts (hunk 1 above)
Comment (cal-smith) [reply src/dialog/dialog.service.ts]: Yeah, in this case it's short enough there's no need to split it up

File: src/dialog/dialog.service.ts [hunk 4]
```diff
@@ -80,59 +107,106 @@ export class DialogService {
...
-			this.dialogRef = null;
-			this.isOpen = false;
-			elementToFocus.focus();
+		// to handle the case where we have a null `this.dialogRef`
+		if (!dialogRef) { return; }
 
-			if (this.dialogSubscription) {
-				this.dialogSubscription.unsubscribe();
-			}
+		let elementToFocus = dialogRef.instance.dialogConfig["previouslyFocusedElement"];
```
[the remark this answers is not in the document]
Comment (cal-smith) [reply src/dialog/dialog.service.ts]: yep 👍

**Pull request merged**